load-testing/tests/relays_stress_helpers_test.go

```diff
@@ -0,0 +1,1143 @@
+
+	gatewaysToStake := gatewayInc
+	if gatewayCount+gatewaysToStake > maxGateways {
+		gatewaysToStake = maxGateways - gatewayCount
+	}
+
+	if gatewaysToStake == 0 {
+		return newGateways
+	}
+
+	logger.Debug().
+		Int64("session_num", sessionInfo.sessionNumber).
+		Int64("block_height", sessionInfo.blockHeight).
+		Msgf(
+			"staking gateways for next session %d (%d->%d)",
+			sessionInfo.sessionNumber+1,
+			gatewayCount,
+			gatewayCount+gatewaysToStake,
+		)
+
+	for gwIdx := int64(0); gwIdx < gatewaysToStake; gwIdx++ {
+		keyName := fmt.Sprintf("gateway%d", gatewayCount+gwIdx+1)
+		gateway := s.addGateway(keyName)
+		s.addPendingStakeGatewayMsg(gateway)
+		s.sendPendingMsgsTx(sessionInfo.blockHeight, gateway)
+		newGateways = append(newGateways, gateway)
+	}
+
+	// The new gateways are returned so the caller can construct delegation messages
+	// given the existing applications.
+	return newGateways
+}
+
+// sendPendingMsgsTx sends a transaction with the provided messages using the keyName provided.
+func (s *relaysSuite) sendPendingMsgsTx(height int64, actor *accountInfo) {
+	// Do not send empty message transactions as trying to do so will make SignTx to fail.
+	if len(actor.pendingMsgs) == 0 {
+		return
+	}
+
+	txBuilder := s.txContext.NewTxBuilder()
+	err := txBuilder.SetMsgs(actor.pendingMsgs...)
+	require.NoError(s, err)
+
+	txBuilder.SetTimeoutHeight(uint64(height + 2))
+	txBuilder.SetGasLimit(690000042)
+
+	// All messages have to be signed by the keyName provided.
+	// TODO_TECHDEBT: Extend the txContext to support multiple signers.
+	err = s.txContext.SignTx(actor.keyName, txBuilder, false, false)
+	if err != nil {
+		require.NoError(s, err)
+	}
+
+	// Serialize transactions.
+	txBz, err := s.txContext.EncodeTx(txBuilder)
+	require.NoError(s, err)
+
+	// Empty the pending messages after the transaction is serialized.
+	actor.pendingMsgs = []sdk.Msg{}
+
+	// txContext.BroadcastTx uses the async mode, if this method changes in the future
+	// to be synchronous, make sure to keep this async to avoid blocking the test.
+	go func() {
+		r, err := s.txContext.BroadcastTx(txBz)
+		require.NoError(s, err)
+		require.NotNil(s, r)
+	}()
+}
+
+// waitForTxsToBeCommitted waits for transactions to observed on-chain.
+// It is used to ensure that the transactions are included in the next block.
+func (s *relaysSuite) waitForTxsToBeCommitted() []*types.TxResult {
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+
+	txResults := []*types.TxResult{}
+	ch := s.newTxEventsObs.Subscribe(ctx).Ch()
+	for {
+		txResult := <-ch
+		txResults = append(txResults, txResult)
+
+		// tm.Event='Tx' sends a transaction event at a time, so the number of transactions
+		// to be observed is unknown.
+		// The number of transactions to be observed is not available in the TxResult
+		// event, so this number is taken from the last block event.
+		var numTxs int
+		// Sometimes the block received from s.latestBlock is the previous one,
```

Comment:
Could you take a second look at the comments in this function. I think it requires additional context on how the internals work and hard for a reader without that level of context.